### ISSUE TITLE
cswarth : xmlrpclib.Fault: <Fault -32700: 'parse error; not well formed'>

### ISSUE THREAD

**Issue by HugoTian** [conda/conda-build#1177]
@cswarth I think this is a conda build question.
Origin link : https://github.com/conda/conda/issues/3234

I keep getting a parse error when using `conda skeleton pypi` on the latest anaconda 2.4.1.
This happens no matter which package I am making a skeleton for.
An older anaconda2 installation using `conda 3.19.0` seems to work fine.

```
$ conda info
Current conda install:

             platform : linux-64
        conda version : 4.1.4
    conda-env version : 2.5.1
  conda-build version : 1.21.3
       python version : 2.7.12.final.0
     requests version : 2.10.0
     root environment : /home/matsengrp/local/anaconda2/2.4  (writable)
  default environment : /home/matsengrp/local/anaconda2/2.4
     envs directories : /home/cwarth/.conda/envs
        package cache : /home/cwarth/.conda/envs/.pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /home/cwarth/.condarc
         offline mode : False
    is foreign system : False

```

```
$ conda skeleton pypi brewer2mpl                                                                                                                                                                                                                                                                                 
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda-build/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/home/matsengrp/local/anaconda2/2.4/bin/conda-skeleton", line 5, in <module>
    sys.exit(main())
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/main_skeleton.py", line 300, in main
    args_func(args, p)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/main_build.py", line 389, in args_func
    args.func(args, p)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/main_skeleton.py", line 315, in execute
    pypi.main(args, parser)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/pypi.py", line 263, in main
    all_packages = [match["name"] for match in client.search({"name": args.packages}, "or")]
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/xmlrpclib.py", line 1243, in __call__
    return self.__send(self.__name, args)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/xmlrpclib.py", line 1602, in __request
    verbose=self.__verbose
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/pypi.py", line 233, in request
    return self.parse_response(resp)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/pypi.py", line 242, in parse_response
    return u.close()
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/xmlrpclib.py", line 800, in close
    raise Fault(**self._stack[0])
xmlrpclib.Fault: <Fault -32700: 'parse error; not well formed'>
```

**Comment by msarahan** [comment]
Confirming that this is a conda-build issue.  PyPI seems to have changed something about its XMLRPC code, and it is breaking all of our skeleton code, which depends on XMLRPC.  We will investigate.  For now, moving this issue to conda-build.

**Comment by cswarth** [comment]
Thanks all for the feedback.  FWIW, I reran the test after updating with `conda install conda-build`, and got a similar error.

```
$ conda info
Current conda install:

             platform : linux-64
        conda version : 4.1.11
    conda-env version : 2.5.2
  conda-build version : 1.21.9
       python version : 2.7.12.final.0
     requests version : 2.10.0
     root environment : /home/matsengrp/local/anaconda2/2.4  (writable)
  default environment : /home/matsengrp/local/anaconda2/2.4
     envs directories : /home/cwarth/.conda/envs
        package cache : /home/cwarth/.conda/envs/.pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /home/cwarth/.condarc
         offline mode : False
    is foreign system : False
```

```
$ conda skeleton pypi brewer2mpl
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): pypi.io
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): pypi.org
Error: XMLRPC Fault while parsing PyPI response.  This is likely a transient error - please try again soon.
```

**Comment by cswarth** [comment]
It seems to have broken between `conda-build=1.20.1` and `conda-build=1.20.2`.
v1.20.1 works fine and v1.20.2 is the first time I see this error,
```
$ conda-build --version
conda-build 1.20.2

$ conda skeleton pypi brewer2mpl                                                                                                                                                                                                                                                                                                                                          
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda-build/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/home/matsengrp/local/anaconda2/2.4/bin/conda-skeleton", line 5, in <module>
    sys.exit(main())
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/main_skeleton.py", line 300, in main
    args_func(args, p)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/main_build.py", line 370, in args_func
    args.func(args, p)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/main_skeleton.py", line 315, in execute
    pypi.main(args, parser)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/site-packages/conda_build/pypi.py", line 269, in main
    all_packages = client.list_packages()
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/xmlrpclib.py", line 1243, in __call__
    return self.__send(self.__name, args)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/xmlrpclib.py", line 1602, in __request
    verbose=self.__verbose
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/xmlrpclib.py", line 1283, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/home/matsengrp/local/anaconda2/2.4/lib/python2.7/xmlrpclib.py", line 1331, in single_request
    response.msg,
xmlrpclib.ProtocolError: <ProtocolError for pypi.io/pypi: 301 Redirect to Primary Domain>



```

**Comment by msarahan** [comment]
Thanks for the help narrowing it down.  At some point, I changed the code from a list-all operation to a search operation to reduce the load on pypi and reduce other sporadic errors - my guess is that this code might be related.

**Comment by cswarth** [comment]
This works,
```
conda skeleton pypi --pypi-url=https://pypi.python.org/pypi brewer2mpl 
```
So I suspect the change make to main_skeleton.py in 2db31bb2c48d24

**Comment by msarahan** [comment]
Yeah, that seems reasonable.  I should just stick with the old URLs for now...  If you want credit for finding this, go ahead and submit a PR.  Otherwise, I'll put this in later today.

Strange that this suddenly started acting up, but so it goes.

**Comment by cswarth** [comment]
I'd probably break something else.  I'll let you get to it when you can.

**Comment by msarahan** [comment]
Fixed in #1181; part of 1.21.11 release.  Packages should be available soon.  Closing this issue.  Please reopen or reraise if issues recur.
